mozjs/src/jsval.rs

```diff
@@ -14,117 +14,126 @@ use crate::jsapi::JS::Symbol;
 use crate::jsapi::JS::TraceKind;
 use crate::jsapi::JS::Value;
 
-use libc::c_void;
 use std::default::Default;
-use std::mem;
+use std::ffi::c_void;
 
 pub type JSVal = Value;
 
 #[cfg(target_pointer_width = "64")]
 const JSVAL_TAG_SHIFT: usize = 47;
 
-#[cfg(target_pointer_width = "64")]
-const JSVAL_TAG_MAX_DOUBLE: u32 = 0x1FFF0u32;
+#[cfg(target_pointer_width = "32")]
+const JSVAL_TAG_SHIFT: usize = 32;
+
+const JSVAL_TAG_MAX_DOUBLE: u32 = 0x1FFF0;
 
 #[cfg(target_pointer_width = "32")]
 const JSVAL_TAG_CLEAR: u32 = 0xFFFFFF80;
 
-#[cfg(target_pointer_width = "64")]
-#[repr(u32)]
-#[allow(dead_code)]
-enum ValueTag {
-    INT32 = JSVAL_TAG_MAX_DOUBLE | (JSValueType::JSVAL_TYPE_INT32 as u32),
-    UNDEFINED = JSVAL_TAG_MAX_DOUBLE | (JSValueType::JSVAL_TYPE_UNDEFINED as u32),
-    STRING = JSVAL_TAG_MAX_DOUBLE | (JSValueType::JSVAL_TYPE_STRING as u32),
-    SYMBOL = JSVAL_TAG_MAX_DOUBLE | (JSValueType::JSVAL_TYPE_SYMBOL as u32),
-    BIGINT = JSVAL_TAG_MAX_DOUBLE | (JSValueType::JSVAL_TYPE_BIGINT as u32),
-    BOOLEAN = JSVAL_TAG_MAX_DOUBLE | (JSValueType::JSVAL_TYPE_BOOLEAN as u32),
-    MAGIC = JSVAL_TAG_MAX_DOUBLE | (JSValueType::JSVAL_TYPE_MAGIC as u32),
-    NULL = JSVAL_TAG_MAX_DOUBLE | (JSValueType::JSVAL_TYPE_NULL as u32),
-    OBJECT = JSVAL_TAG_MAX_DOUBLE | (JSValueType::JSVAL_TYPE_OBJECT as u32),
+const CELL_ALIGN_MASK: u64 = (1 << 3) - 1;
+
+macro_rules! value_tag {
+    (enum $name:ident { $($variant:ident = $value_type:path,)* }) => {
+        #[cfg(target_pointer_width = "64")]
+        #[repr(u32)]
+        #[allow(dead_code)]
+        enum $name {
+            $(
+                $variant = JSVAL_TAG_MAX_DOUBLE | ($value_type as u32),
+            )*
+        }
+
+        #[cfg(target_pointer_width = "32")]
+        #[repr(u32)]
+        #[allow(dead_code)]
+        enum $name {
+            $(
+                $variant = JSVAL_TAG_CLEAR | ($value_type as u32),
+            )*
+        }
+    };
```

Comment:
Sorry, drive by review:

I feel like the addition of a macro here obfuscates this code a bit and the definition of `ValueTag` seems identical between the two pointer sizes apart from the clear value.. Would it be possible to instead conditionally define a value to bitwise or with the ValueTag members?

```rust
#[cfg(target_pointer_width = "32")]
const JSVAL_TAG_CLEAR: u32 = 0xFFFFFF80;
#[cfg(target_pointer_width = "64")]
const JSVAL_TAG_CLEAR: u32 = 0x1FFF0;

#[repr(u32)]
#[allow(dead_code)]
enum ValueTag {
   INT32 = JSVAL_TAG_CLEAR | (JSValueType::JSVAL_TYPE_INT32 as u32),
    UNDEFINED = JSVAL_TAG_CLEAR | (JSValueType::JSVAL_TYPE_UNDEFINED as u32),
    STRING = JSVAL_TAG_CLEAR | (JSValueType::JSVAL_TYPE_STRING as u32),
    SYMBOL = JSVAL_TAG_CLEAR | (JSValueType::JSVAL_TYPE_SYMBOL as u32),
    BIGINT = JSVAL_TAG_CLEAR | (JSValueType::JSVAL_TYPE_BIGINT as u32),
    BOOLEAN = JSVAL_TAG_CLEAR | (JSValueType::JSVAL_TYPE_BOOLEAN as u32),
    MAGIC = JSVAL_TAG_CLEAR | (JSValueType::JSVAL_TYPE_MAGIC as u32),
    NULL = JSVAL_TAG_CLEAR | (JSValueType::JSVAL_TYPE_NULL as u32),
    OBJECT = JSVAL_TAG_CLEAR | (JSValueType::JSVAL_TYPE_OBJECT as u32),
}
```

Reply:
I see what you mean. I don't think the 64-bit one would count as a "clear tag" though.
Looking at [Value.h#207](https://searchfox.org/mozilla-central/source/js/public/Value.h#207) and [Value.h#230](https://searchfox.org/mozilla-central/source/js/public/Value.h#230), they are specifically separate.
I was trying to make the implementation align with the C++ one, so I left them separate. They also seem to serve different purposes in terms of the bit representation.

Reply:
Hrm. If they should be represented separately, I don't think sharing the macro is the best way. Maybe instead come up with a good name for the thing you bitwise or with and add a comment explaining what is going on here.